src/providers/trusted_service/context/key_management.rs

```diff
@@ -0,0 +1,146 @@
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        )?;
+        let handle = proto_resp.handle;
+
+        let proto_req = Box::new(DestroyKeyIn { handle });
+        let mut proto_resp = Box::new(DestroyKeyOut::default());
+        Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::DestroyKey,
+            self.rpc_caller,
+        )?;
+        Ok(())
+    }
+
+    pub fn check_key_exists(&self, id: u32) -> Result<bool, Error> {
+        let proto_req = Box::new(OpenKeyIn { id });
+        let mut proto_resp = Box::new(OpenKeyOut::default());
+        match Context::send_request(
+            &proto_req,
+            &mut proto_resp,
+            Opcode::OpenKey,
+            self.rpc_caller,
+        ) {
+            Ok(_) => {
+                let handle = proto_resp.handle;
+                let proto_req = Box::new(CloseKeyIn { handle });
+                Context::send_request(
+                    &proto_req,
+                    &mut proto_resp,
+                    Opcode::OpenKey,
+                    self.rpc_caller,
+                )?;
+                Ok(true)
+            }
+            Err(e) => {
+                if e == Error::DoesNotExist {
+                    Ok(false)
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
+
+    fn send_request<Req: Message, Resp: Message>(
+        req: &Box<Req>,
+        resp: &mut Box<Resp>,
+        opcode: Opcode,
+        rpc_cl: *mut rpc_caller,
+    ) -> Result<(), Error> {
+        info!("Beginning call to Trusted Service");
+        let mut buf_out = null_mut();
+
+        let call_handle = unsafe { rpc_caller_begin(rpc_cl, &mut buf_out, req.encoded_len()) };
+        if call_handle == null_mut() {
+            error!("Call handle was null");
+            return Err(Error::CommunicationFailure);
+        }
+        let mut buf_out = unsafe { Vec::from_raw_parts(buf_out, 0, req.encoded_len()) };
+        req.encode(&mut buf_out).unwrap();
+        forget(buf_out);
+
+        println!("Invoking call");
+        let mut opstatus = 0;
+        let mut buf_in = null_mut();
+        let mut buf_in_size = 0;
+        let status = unsafe {
+            rpc_caller_invoke(
+                rpc_cl,
+                call_handle,
+                i32::from(opcode).try_into().unwrap(),
+                &mut opstatus,
+                &mut buf_in,
+                &mut buf_in_size,
+            )
+        };
+        if status != 0 || opstatus != 0 {
+            unsafe { rpc_caller_end(rpc_cl, call_handle) };
+            error!("Error: status = {}, opstatus = {}", status, opstatus as i32);
+            return Status::from(opstatus as i32).to_result();
+        }
+        let buf_in = unsafe { Vec::from_raw_parts(buf_in, 0, buf_in_size) };
+        resp.merge(&*buf_in).unwrap();
```

Comment:
ah ok here we write in `resp` the encoded content in `buf_in`. Could we return `resp` to make it clear this is the result?

Reply:
I'll try, but I'm not 100% sure it's possible

Reply:
Ok, made it work, it's actually nicer to work with and follows the "don't pass return arguments as params" guideline 😄